### PR TITLE
IAST Overhead control engine

### PR DESCRIPTION
**Pull request by CarlesDD**
### What does this PR do?
Implement Overhead Control Engine (OCE) for IAST operations

### Motivation
Low overhead is a mandatory requirement for a software that is running in production. The main constraint is that having to introduce many hooks for taint tracking keeping a low overhead is a challenging task.

The OCE is an element that by design ensures that the overhead does not go over a maximum limit. It will measure operations being executed in a request and it will deactivate detection (and therefore reduce the overhead to nearly 0) if a certain threshold is reached.

### Plugin Checklist

- [ ] Unit tests.

### CODE REVIEW

File: packages/dd-trace/src/appsec/iast/overhead-controller.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  return operation.hasQuota(iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY])
+}
+
+function hasQuotaLongRunning (operation, iastContext) {
+  iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY] = iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY] || _initializeContext()
+  const reserved = operation.hasQuota(iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY])
+  return new Quota(reserved, () => operation.release(iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY]))
+}
+
+function _initializeContext () {
```
Comment (uurien) [inline packages/dd-trace/src/appsec/iast/overhead-controller.js:73]: If we remove this method, and add a method ilke `initInRequest` in `SINGLE_SHOT_OPERATIONS.REPORT_VULNERABILITY` we will be able to init all automatically when the request starts.
At this way, we will define each OPERATION in the same point.

```
REPORT_VULNERABILITY: {
    hasQuota: (context) => {
      if (!context.isRequestAnalyzed) return false
      const reserved = !!context.tokens[REPORT_VULNERABILITY] && context.tokens[REPORT_VULNERABILITY] > 0
      if (reserved) {
        context.tokens[REPORT_VULNERABILITY]--
      }
      return reserved
    },
    name: REPORT_VULNERABILITY,
    initialTokenBucketSize: 2,
    initInRequest () {
        context.tokens[REPORT_VULNERABILITY] = SINGLE_SHOT_OPERATIONS[REPORT_VULNERABILITY].initialTokenBucketSize
    }
  }
```

File: packages/dd-trace/test/appsec/iast/overhead-controller.spec.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    it('should populate oce context', () => {
+      const iastContext = {}
+      overheadController.initializeRequestContext(iastContext)
+      expect(iastContext).to.have.nested.property(overheadController.OVERHEAD_CONTROLLER_CONTEXT_KEY)
+    })
+  })
+
+  describe('Single shot operation', () => {
+    it('should not allow with no context', () => {
+      expect(overheadController.hasQuotaSingleShot(overheadController.SINGLE_SHOT_OPERATIONS[0])).to.be.false
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/overhead-controller.spec.js:16]: Is not `SINGLE_SHOT_OPERATIONS[0]` undefined?
Maybe you want to use `SINGLE_SHOT_OPERATIONS. REPORT_VULNERABILITY`

File: packages/dd-trace/test/appsec/iast/overhead-controller.spec.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+      const SINGLE_SHOT_OPERATION = overheadController.SINGLE_SHOT_OPERATIONS.REPORT_VULNERABILITY
+
+      beforeEach(() => {
+        iastContext = {}
+        overheadController.initializeRequestContext(iastContext)
+      })
+
+      it('should populate initial context with available tokens', () => {
+        const _iastContext = {}
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/overhead-controller.spec.js:32]: you have already defined `let iastContext` in parent scope, why are you creating this new `_iastContext` if this test is not using the original one?

File: packages/dd-trace/src/appsec/iast/overhead-controller.js [hunk 4]
```diff
@@ -65,7 +65,8 @@ const SINGLE_SHOT_OPERATIONS = {
 }
 
 function hasQuotaSingleShot (operation, iastContext) {
-  if (!iastContext || !iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY]) {
+  const oceContext = iastContext && iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY]
+  if (!oceContext) {
     return false
   }
   return operation.hasQuota(iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY])
```
Comment (uurien) [inline packages/dd-trace/src/appsec/iast/overhead-controller.js:72]: You have create a var `oceContext` to use at this point and prevent the access to the object property

File: packages/dd-trace/src/appsec/iast/overhead-controller.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      this._operationRelease()
+      this._released = true
+    }
+  }
+}
+
+const LONG_RUNNING_OPERATIONS = {
+  ANALYZE_REQUEST: {
+    hasQuota: () => {
+      const reserved = concurrentRequestTokens > 0
```
Comment (manuel-alvarez-alvarez) [inline packages/dd-trace/src/appsec/iast/overhead-controller.js:30]: The initial version was targeting just a percentage the requests and then we wanted to move to the concurrent number of requests. In the end, the important thing is that the API is designed in a way that we can plug different strategies, for example:
1. Analyze at most two concurrent requests 
2. Analyze only 30% of the requests
3. Analyze each endpoint only once
4. Allow only two concurrent requests and only 30% of the requests
5. ...

**Review by iunanua — APPROVED** [review]

**Review by hoolioh — APPROVED** [review]

File: packages/dd-trace/src/appsec/iast/overhead-controller.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+function _resetGlobalContext () {
+  Object.assign(GLOBAL_OCE_CONTEXT, _getNewContext())
+}
+
+function startGlobalContextResetInterval () {
+  if (!globalContextResetInterval) {
+    globalContextResetInterval = setInterval(() => {
+      _resetGlobalContext()
+    }, GLOBAL_CONTEXT_RESET_INTERVAL)
```
Comment (uurien) [inline packages/dd-trace/src/appsec/iast/overhead-controller.js:68]: Each time that we call to `setInterval` or `setTimeout` we could call to `.unref()` method: https://nodejs.org/api/timers.html#timeoutrefresh
```suggestion
    }, GLOBAL_CONTEXT_RESET_INTERVAL)
    globalContextResetInterval.unref()
```

File: packages/dd-trace/src/appsec/iast/overhead-controller.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+function _getContext (iastContext) {
+  if (iastContext && iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY]) {
+    return iastContext[OVERHEAD_CONTROLLER_CONTEXT_KEY]
+  }
+  return GLOBAL_OCE_CONTEXT
+}
+
+function acquireRequest (rootSpan) {
+  return (rootSpan.context().toSpanId() % 3) === 0
```
Comment (uurien) [inline packages/dd-trace/src/appsec/iast/overhead-controller.js:48]: Are you sure that `toSpanId()` returns a number? I've seen that in the implementation:
```javascript
  toSpanId () {
    return this._spanId.toString(10)
  }
```

File: packages/dd-trace/test/appsec/iast/overhead-controller.spec.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+          spanId: 3
+        }))
+      }
+
+      const reserved = overheadController.acquireRequest(rootSpan)
+      expect(reserved).to.be.true
+    })
+  })
+
+  it('should not allow requests which span id is not divisible by 3', () => {
```
Comment (uurien) [inline packages/dd-trace/test/appsec/iast/overhead-controller.spec.js:54]: Isn't this test part of `Analyze request`? It is out of the `describe('Analyze request'`

File: packages/dd-trace/src/appsec/iast/overhead-controller.js (hunk 7 above)
Comment (uurien) [reply packages/dd-trace/src/appsec/iast/overhead-controller.js:48]: Forget this comment, `'2' % 3 === 2`

File: packages/dd-trace/src/appsec/iast/overhead-controller.js (hunk 6 above)
Comment (uurien) [reply packages/dd-trace/src/appsec/iast/overhead-controller.js:68]: Nice, I didn't know that it exists, thanks

**Pull request merged**